### PR TITLE
Add `union_extract` kernel

### LINKED ISSUE #6386: Add union_extract kernel
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Retrieve the value of the given union variant, or NULL if it's not currently selected
This is will be used at https://github.com/apache/datafusion/issues/11081

**Describe the solution you'd like**

A kernel in `arrow_select` with the following signature:
 `fn union_extract(union_array: &UnionArray, field: &str) -> Result<ArrayRef>`

**Describe alternatives you've considered**

Implement this in datafusion, but after comments from @samuelcolvin and @alamb at https://github.com/apache/arrow-rs/issues/6247#issuecomment-2307370777 and https://github.com/apache/arrow-rs/issues/6247#issuecomment-2307418058, I decided to do here

**Additional context**

### PR DESCRIPTION
**Pull request by gstvg**
# Which issue does this PR close?

Closes #6386 

# Rationale for this change

# What changes are included in this PR?

`union_extract` kernel implementation, docs and tests

# Are there any user-facing changes?

A new public kernel `union_extract` is added

### CODE REVIEW

**Comment by alamb** [comment]
I am depressed about the large review backlog in this crate. We are looking for more help from the community reviewing PRs -- see https://github.com/apache/arrow-rs/issues/6418 for more

**Comment by alamb** [comment]
I am depressed about the large review backlog in this crate. We are looking for more help from the community reviewing PRs -- see https://github.com/apache/arrow-rs/issues/6418 for more

**Comment by wiedld** [comment]
I've started reviewing this one, but it's going to take me a bit (learning) to make it thru. Thank you for the documentation and testing.

File: arrow-select/src/union_extract.rs [hunk 1]
```diff
@@ -0,0 +1,1221 @@
...
+    if offsets.is_empty() {
+        return true;
+    }
+
+    // checks a common form of non sequential offsets, when sequential nulls reuses the same value,
+    // pointed by the same offset, while valid values offsets increases one by one
+    // this also checks if the last chunk/remainder is sequential relative to the first offset
+    if offsets[0] + offsets.len() as i32 - 1 != offsets[offsets.len() - 1] {
+        return false;
+    }
```
Comment (wiedld) [inline arrow-select/src/union_extract.rs:365]: I didn't quite get the first explanation. 

But the second reason (that it checks the remainder when we have a single remainder element past the last chunk boundary) makes sense -- and is covered in the tests. 👍🏼

File: arrow-select/src/union_extract.rs [hunk 2]
```diff
@@ -0,0 +1,1221 @@
...
+                eq_scalar(&values, 2),
+                BoolValue::Buffer(cross_check(&values, 2))
+            );
+
+            values[i] = 1;
+        }
+    }
+
+    #[test]
+    fn test_is_sequential() {
```
Comment (wiedld) [inline arrow-select/src/union_extract.rs:451]: Can we also add `assert!(!is_sequential(&[1, 2, 3, 5, 6, 7]));` to demonstrate that it checks the incr at the chunk boundary?

**Review by wiedld — COMMENTED** [review]
Thank you for the well documented code.

File: arrow-select/src/union_extract.rs [hunk 3]
```diff
@@ -0,0 +1,1221 @@
...
+        .iter()
+        .copied()
+        .enumerate()
+        .fold(true, |acc, (i, offset)| {
+            acc & (offset == remainder[0] + i as i32)
+        }) //if the remainder is sequential relative to the first offset is checked at the start of the function
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (wiedld) [inline arrow-select/src/union_extract.rs:393]: Also confirmed branch test coverage with [cov](https://docs.rs/cov/0.0.5/cov/). 👍🏼

File: arrow-select/src/union_extract.rs [hunk 4]
```diff
@@ -0,0 +1,1221 @@
...
+        }
+    } else {
+        (true_bits, true)
+    };
+
+    // restrict to chunk boundaries
+    let set_bits = set_bits - set_bits % 64;
+
+    let mut buffer =
+        MutableBuffer::new(bit_util::ceil(type_ids.len(), 64) * 8).with_bitset(set_bits / 8, val);
```
[the remark this answers is not in the document]
Comment (gstvg) [reply arrow-select/src/union_extract.rs]: I'm not sure relative to which part, will explain both:

Without `set_bits = set_bits - set_bits % 64`, if the `type_ids` len is already a multiple of 64 and aligned, and `set_bits` is not multiple of 64, `buffer.extend` below would push an additional partial packed mask, requiring a buffer grow. Even with `EQ_SCALAR_CHUNK_SIZE = 512`, `set_bits` can be a non multiple of 64 if the first bit flip happens in the last chunk, which is potentially smaller than 512

Relative to `bit_util::ceil(type_ids.len(), 64) * 8`, I didn't realized that in `MutableBuffer::new/with_capacity`, so yes, using 64 as divisor and then `* 8` is redundant, removed it. Thanks!

I also noticed that `buffer.truncate` is unnecessary so I removed it too. `BooleanBuffer::new` uses the correct len from `type_ids`

File: arrow-select/src/union_extract.rs [hunk 5]
```diff
@@ -0,0 +1,1221 @@
...
+    }
+}
+
+const EQ_SCALAR_CHUNK_SIZE: usize = 512;
+
+#[derive(Debug, PartialEq)]
+enum BoolValue {
+    Scalar(bool),
+    Buffer(BooleanBuffer),
+}
```
[the remark this answers is not in the document]
Comment (gstvg) [reply arrow-select/src/union_extract.rs]: Done

File: arrow-select/src/union_extract.rs (hunk 1 above)
Comment (gstvg) [reply arrow-select/src/union_extract.rs]: I updated the comment to more clear and added an example

This is used on [C++](https://github.com/apache/arrow/blob/81b94dcda6180781effc331650d9cd413a4ac115/cpp/src/arrow/array/builder_union.h#L125) and [Go](https://github.com/apache/arrow/blob/37f62d0bc5f4d22e7194947963b445225b984558/go/arrow/array/union.go#L1173) union builders
Our `UnionBuilder` doesn't have an equivalent `append_nulls` method

File: arrow-select/src/union_extract.rs (hunk 2 above)
Comment (gstvg) [reply arrow-select/src/union_extract.rs:459]: Good catch! Added with a few more variations too.

**Comment by gstvg** [comment]
> Thank you for the well documented code.

I pushed like dozens of faster ways of doing the same thing, felt obligated to document it :sweat_smile: 

Thanks @wiedld for your review

**Review by alamb — APPROVED** [review]
Thank you @gstvg  and @wiedld  for the review. This looks 💯  to me ❤️

**Comment by alamb** [comment]
I'll plan to merge this tomorrow unless there are any other comments or anyone else wants to comment

**Comment by alamb** [comment]
Looks like the CI failure is not related -- https://github.com/apache/arrow-rs/actions/runs/10984024899/job/30494124939?pr=6387

**Comment by alamb** [comment]
I restarted the failed CI job https://github.com/apache/arrow-rs/actions/runs/10984024899/job/30494124939?pr=6387

I think it will pass now that https://github.com/apache/arrow-rs/pull/6437 is merged

**Comment by alamb** [comment]
I took the liberty of merging up from `master` to try and get a clean CI run for this PR

**Comment by alamb** [comment]
This CI run looks entirely unrelated - I will close/reopen this PR to retrigger and see if it is still happening

**Comment by alamb** [comment]
integration CI is failing due to https://github.com/apache/arrow-rs/issues/6448

**Comment by alamb** [comment]
https://github.com/apache/arrow-rs/issues/6448 is fixed, so merging it in

**Comment by alamb** [comment]
Thanks again @gstvg  and @wiedld

**Pull request merged**